invalid_auth on some Slack workspaces

**Describe the bug**
Using enterprise/paid version of Slack, after configuring a secrets.txt file, running slackdump -u, it is unable to download the users or anything else:
2022/03/16 20:19:25 > checking user cache...
2022/03/16 20:19:25   caching users for the first time
2022/03/16 20:19:25 error fetching users: couldn't fetch users

verbose output from slackdump -v -u:
2022/03/16 20:22:22 slackdump.go:88: > checking user cache...
2022/03/16 20:22:22 users.go:37:   caching users for the first time
2022/03/16 20:22:22 main.go:65: error fetching users: couldn't fetch users

**To Reproduce**
Steps to reproduce the behavior:
When I used chrome to get the value of the authorization cookie, I did have to do URL decode so that it would appear as a base64 value.   It does not appear that it is complaining about authorization.

**Expected behavior**
A clear and concise description of what you expected to happen.
Expected it to list the users

**Output**
If applicable, add output to help explain your problem.
Output is shown above.

**Desktop (please complete the following information):**
 - OS: [e.g. macOS]
 Windows 11 using released binaries.

**Additional context**
Add any other context about the problem here.

--- a/cmd/slackdump/main.go
+++ b/cmd/slackdump/main.go
@@ -146,7 +146,7 @@ func parseCmdLine(args []string) (params, error) {
 
 	// authentication
 	fs.StringVar(&p.appCfg.Creds.Token, "t", os.Getenv(slackTokenEnv), "Specify slack `API_token`, (environment: "+slackTokenEnv+")")
-	fs.StringVar(&p.appCfg.Creds.Cookie, "cookie", os.Getenv(slackCookieEnv), "d= cookie `value` (environment: "+slackCookieEnv+")")
+	fs.StringVar(&p.appCfg.Creds.Cookie, "cookie", os.Getenv(slackCookieEnv), "d= cookie `value` or a path to a cookie.txt file (environment: "+slackCookieEnv+")")
 
 	// operation mode
 	fs.BoolVar(&p.appCfg.ListFlags.Channels, "c", false, "same as -list-channels")

--- a/go.mod
+++ b/go.mod
@@ -13,10 +13,11 @@ require (
 )
 
 require (
+	github.com/MercuryEngineering/CookieMonster v0.0.0-20180304172713-1584578b3403 // indirect
 	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/gorilla/websocket v1.4.2 // indirect
 	github.com/pmezard/go-difflib v1.0.0 // indirect
 	gopkg.in/yaml.v2 v2.2.2 // indirect
 )
 
-replace github.com/slack-go/slack => github.com/rusq/slack v0.10.2
+replace github.com/slack-go/slack => github.com/rusq/slack v0.10.3

--- a/go.mod
+++ b/go.mod
@@ -3,6 +3,7 @@ module github.com/rusq/slackdump
 go 1.17
 
 require (
+	github.com/MercuryEngineering/CookieMonster v0.0.0-20180304172713-1584578b3403
 	github.com/golang/mock v1.6.0
 	github.com/joho/godotenv v1.4.0
 	github.com/pkg/errors v0.9.1
@@ -13,7 +14,6 @@ require (
 )
 
 require (
-	github.com/MercuryEngineering/CookieMonster v0.0.0-20180304172713-1584578b3403 // indirect
 	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/gorilla/websocket v1.4.2 // indirect
 	github.com/pmezard/go-difflib v1.0.0 // indirect

--- a/internal/app/config.go
+++ b/internal/app/config.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"html/template"
 	"io"
+	"os"
 	"strings"
 
 	"github.com/rusq/slackdump"
@@ -84,8 +85,34 @@ type SlackCreds struct {
 	Cookie string
 }
 
-func (c SlackCreds) Valid() bool {
-	return c.Token != "" && c.Cookie != ""
+func (c SlackCreds) Validate() error {
+	if c.Token == "" {
+		return fmt.Errorf("token is not specified")
+	} else if c.Cookie == "" {
+		return fmt.Errorf("cookie is not specified")
+	}
+
+	stat, err := os.Stat(c.Cookie)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// cookie is not a file
+			return nil
+		}
+		return fmt.Errorf("cookie file: stat error: %w", err)
+	}
+
+	if stat.IsDir() {
+		return fmt.Errorf("cookie path is a directory")
+	}
+
+	// cookie is a file, try to open it for reading
+	f, err := os.Open(c.Cookie)
+	if err != nil {
+		return fmt.Errorf("error opening the cookie file: %w", err)
+	}
+	f.Close()
+
+	return nil
 }
 
 type ListFlags struct {
@@ -99,8 +126,8 @@ func (lf ListFlags) FlagsPresent() bool {
 
 // Validate checks if the command line parameters have valid values.
 func (p *Config) Validate() error {
-	if !p.Creds.Valid() {
-		return fmt.Errorf("slack token or cookie not specified")
+	if err := p.Creds.Validate(); err != nil {
+		return err
 	}
 
 	if !p.Input.IsValid() && !p.ListFlags.FlagsPresent() {

--- a/internal/app/config.go
+++ b/internal/app/config.go
@@ -74,10 +74,14 @@ func (out Output) IsText() bool {
 	return out.Format == OutputTypeText
 }
 
+// SlackCreds stores Slack credentials.
 type SlackCreds struct {
-	Token         string
-	Cookie        string
-	SSBInstanceID string
+	// Token is the Slack auth-token value.
+	Token string
+	// Cookie may contain cookie value or a filename.  I have mixed filings
+	// about this, but blame CURL, it's the same approach, and I have no
+	// imagination.
+	Cookie string
 }
 
 func (c SlackCreds) Valid() bool {

--- a/internal/app/config.go
+++ b/internal/app/config.go
@@ -79,9 +79,11 @@ func (out Output) IsText() bool {
 type SlackCreds struct {
 	// Token is the Slack auth-token value.
 	Token string
-	// Cookie may contain cookie value or a filename.  I have mixed filings
-	// about this, but blame CURL, it's the same approach, and I have no
-	// imagination.
+	// Cookie may contain cookie value or a filename.  I have mixed feelings
+	// about this, but blame cURL, they use the same approach for --cookie, and
+	// I have no imagination.  Also, it's simpler from the usability POV, as
+	// there's no additional parameter.  Well, I'll repeat this to myself until
+	// I start to believe in it.
 	Cookie string
 }
 

--- a/internal/app/config_test.go
+++ b/internal/app/config_test.go
@@ -1,6 +1,8 @@
 package app
 
 import (
+	"os"
+	"path/filepath"
 	"testing"
 
 	"github.com/rusq/slackdump"
@@ -77,6 +79,74 @@ func TestConfig_compileValidateTemplate(t *testing.T) {
 			}
 			if err := p.compileValidateTemplate(); (err != nil) != tt.wantErr {
 				t.Errorf("Config.compileValidateTemplate() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func TestSlackCreds_Validate(t *testing.T) {
+	// prepare environment
+	tempdir := t.TempDir()
+
+	notafile := filepath.Join(tempdir, "fakecookie.txt")
+	if err := os.Mkdir(notafile, 0700); err != nil {
+		t.Fatal(err)
+	}
+
+	testfile := filepath.Join(tempdir, "cookie.txt")
+	if err := os.WriteFile(testfile, []byte("xxx"), 0600); err != nil {
+		t.Fatal(err)
+	}
+	// end prep
+
+	type fields struct {
+		Token  string
+		Cookie string
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		wantErr bool
+	}{
+		{
+			"empty",
+			fields{},
+			true,
+		},
+		{
+			"token missing",
+			fields{Cookie: "$hey"},
+			true,
+		},
+		{
+			"cookie missing",
+			fields{Token: "$hey"},
+			true,
+		},
+		{
+			"all ok, cookie is not a file",
+			fields{Token: "$tok", Cookie: "$hey"},
+			false,
+		},
+		{
+			"all ok, cookie is a file",
+			fields{Token: "$tok", Cookie: testfile},
+			false,
+		},
+		{
+			"all ok, cookie is a fs object, but not a file",
+			fields{Token: "$tok", Cookie: notafile},
+			true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := SlackCreds{
+				Token:  tt.fields.Token,
+				Cookie: tt.fields.Cookie,
+			}
+			if err := c.Validate(); (err != nil) != tt.wantErr {
+				t.Errorf("SlackCreds.Validate() error = %v, wantErr %v", err, tt.wantErr)
 			}
 		})
 	}

--- a/slackdump.go
+++ b/slackdump.go
@@ -9,6 +9,7 @@ import (
 	"time"
 	"unicode/utf8"
 
+	cookiemonster "github.com/MercuryEngineering/CookieMonster"
 	"github.com/pkg/errors"
 	"github.com/rusq/dlog"
 	"github.com/slack-go/slack"
@@ -76,11 +77,25 @@ func New(ctx context.Context, token string, cookie string, opts ...Option) (*Sla
 }
 
 func NewWithOptions(ctx context.Context, token string, cookie string, opts Options) (*SlackDumper, error) {
+	var sopts []slack.Option
+	if isExistingFile(cookie) {
+		dlog.Debug("cookie value appears to be an existing file")
+		cookies, err := cookiemonster.ParseFile(cookie)
+		if err != nil {
+			return nil, fmt.Errorf("error loading cookies file: %w", errors.WithStack(err))
+		}
+		sopts = append(sopts, slack.OptionCookieRAW(cookies...))
+	} else {
+		sopts = []slack.Option{
+			slack.OptionAuthCookie(cookie),
+			slack.OptionCookie("d-s", fmt.Sprintf("%d", time.Now().Unix()-10)),
+		}
+	}
+
 	sd := &SlackDumper{
 		client: slack.New(
 			token,
-			slack.OptionAuthCookie(cookie),
-			slack.OptionCookie("d-s", fmt.Sprintf("%d", time.Now().Unix()-10)),
+			sopts...,
 		),
 		options: opts,
 	}
@@ -99,6 +114,11 @@ func NewWithOptions(ctx context.Context, token string, cookie string, opts Optio
 
 func (sd *SlackDumper) limiter(t tier) *rate.Limiter {
 	return newLimiter(t, sd.options.Tier3Burst, int(sd.options.Tier3Boost))
+}
+
+func isExistingFile(cookie string) bool {
+	_, err := os.Stat(cookie)
+	return err == nil
 }
 
 var ErrRetryFailed = errors.New("callback was not able to complete without errors within the allowed number of retries")

--- a/slackdump.go
+++ b/slackdump.go
@@ -76,20 +76,25 @@ func New(ctx context.Context, token string, cookie string, opts ...Option) (*Sla
 	return NewWithOptions(ctx, token, cookie, options)
 }
 
-func NewWithOptions(ctx context.Context, token string, cookie string, opts Options) (*SlackDumper, error) {
-	var sopts []slack.Option
-	if isExistingFile(cookie) {
-		dlog.Debug("cookie value appears to be an existing file")
-		cookies, err := cookiemonster.ParseFile(cookie)
-		if err != nil {
-			return nil, fmt.Errorf("error loading cookies file: %w", errors.WithStack(err))
-		}
-		sopts = append(sopts, slack.OptionCookieRAW(cookies...))
-	} else {
-		sopts = []slack.Option{
+func makeSlakeOpts(cookie string) ([]slack.Option, error) {
+	if !isExistingFile(cookie) {
+		return []slack.Option{
 			slack.OptionAuthCookie(cookie),
 			slack.OptionCookie("d-s", fmt.Sprintf("%d", time.Now().Unix()-10)),
-		}
+		}, nil
+	}
+	dlog.Debug("cookie value appears to be an existing file")
+	cookies, err := cookiemonster.ParseFile(cookie)
+	if err != nil {
+		return nil, fmt.Errorf("error loading cookies file: %w", errors.WithStack(err))
+	}
+	return []slack.Option{slack.OptionCookieRAW(cookies...)}, nil
+}
+
+func NewWithOptions(ctx context.Context, token string, cookie string, opts Options) (*SlackDumper, error) {
+	sopts, err := makeSlakeOpts(cookie)
+	if err != nil {
+		return nil, err
 	}
 
 	sd := &SlackDumper{
@@ -117,8 +122,8 @@ func (sd *SlackDumper) limiter(t tier) *rate.Limiter {
 }
 
 func isExistingFile(cookie string) bool {
-	_, err := os.Stat(cookie)
-	return err == nil
+	fi, err := os.Stat(cookie)
+	return err == nil && !fi.IsDir()
 }
 
 var ErrRetryFailed = errors.New("callback was not able to complete without errors within the allowed number of retries")

--- a/slackdump_test.go
+++ b/slackdump_test.go
@@ -3,6 +3,8 @@ package slackdump
 import (
 	"context"
 	"math"
+	"os"
+	"path/filepath"
 	"testing"
 	"time"
 
@@ -324,6 +326,32 @@ func Test_newLimiter(t *testing.T) {
 
 			assert.NoError(t, err)
 			assert.WithinDurationf(t, start.Add(tt.wantDelay), stop, 10*time.Millisecond, "delayed for: %s, expected: %s", stop.Sub(start), tt.wantDelay)
+		})
+	}
+}
+
+func Test_isExistingFile(t *testing.T) {
+	testfile := filepath.Join(t.TempDir(), "cookies.txt")
+	if err := os.WriteFile(testfile, []byte("blah"), 0600); err != nil {
+		t.Fatal(err)
+	}
+
+	type args struct {
+		cookie string
+	}
+	tests := []struct {
+		name string
+		args args
+		want bool
+	}{
+		{"not a file", args{"$blah"}, false},
+		{"file", args{testfile}, true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := isExistingFile(tt.args.cookie); got != tt.want {
+				t.Errorf("isExistingFile() = %v, want %v", got, tt.want)
+			}
 		})
 	}
 }

